README.md
# [Flask DataTables Sample](https://blog.appseed.us/flask-data-tables-open-source-sample/)

Open-source sample provided by [AppSeed](https://appseed.us). The project implements [paginated access to data using Flask](https://blog.appseed.us/flask-data-tables-open-source-sample/) and **Simple-DataTables**, a lightweight, extendable, dependency-free javascript HTML table plugin (no jQuery dependency).

<br />

> Features:

- DataTables managed by `Simple-DataTables`  (Vanilla) JS
- Stack: Flask, SqlAlchemy, Flask-Migrate, Flask-RestX
- Data Tables Implementation(s):
  - Loaded from `Data` table by a controller (route)
  - Served by `/api/data` API node and consumed from JS
  - Loaded without any processing from a file:
    - `app/static/datatables/data.json`
  - Inline Edit / Delete
- UI Kit: **Volt Dashboard** (Free Version) by **Themesberg**
- Deployment scripts: Docker, Gunicorn/Nginx, HEROKU
- Support via **Github** (issues tracker) and [Discord](https://discord.gg/fZC6hup).

<br />

> Links:

- 👉 [Flask DataTables Sample](#) - LIVE Demo (coming soon)
- 👉 More [Free Samples](https://appseed.us/admin-dashboards/open-source) - provided by AppSeed

<br />

## ✨ Quick Start in `Docker`

> Get the code

```bash
$ git clone https://github.com/app-generator/flask-volt-datatables.git
$ cd flask-volt-datatables
```

> Start the app in Docker

```bash
$ docker-compose up --build 
```

Visit `http://localhost:85` in your browser. The app should be up & running.

<br />

![Flask DataTables Sample - Open-Source Sample Project provided by AppSeed.](https://user-images.githubusercontent.com/51070104/152824173-a55c9ddd-e282-4dac-bd59-07b27c41269a.gif)

<br />

## ✨ How to use it

> Clone Sources (this repo)

```bash
$ git clone https://github.com/app-generator/flask-volt-datatables.git
$ cd flask-volt-datatables
```

<br />

> Install Modules using a Virtual Environment

```bash
$ virtualenv env
$ source env/bin/activate
$ pip3 install -r requirements.txt
```

Or for **Windows-based Systems**

```bash
$ virtualenv env
$ .\env\Scripts\activate
$
$ # Install modules - SQLite Database
$ pip3 install -r requirements.txt
```

<br />

> Set up the environment

```bash
$ export FLASK_APP=run.py
$ export FLASK_ENV=development
```

Or for **Windows-based Systems**

```bash
$ # CMD terminal
$ set FLASK_APP=run.py
$ set FLASK_ENV=development
$
$ # Powershell
$ $env:FLASK_APP = ".\run.py"
$ $env:FLASK_ENV = "development"
```

<br />

> Load Sample Data `media/data.csv`

```bash
$ flask load_data  # randomize the timestamp
// OR
$ flask load_random_data  # randomize the timestamp and values
```

<br />

> Start the APP

```bash
$ flask run 
```

The paginated information is available in three ways: 

- Loaded from `Data` table by a controller (route)
- Served by `/api/data` API node and consumed from JS
- Loaded without any processing from a file:
  - `app/static/datatables/data.json`  

<br />
## Code-base structure

The project has a simple structure, represented as bellow:

```bash
< PROJECT ROOT >
   |
   |-- app/__init__.py
   |-- app/
   |    |-- static/
   |    |    |-- <css, JS, images>         # CSS files, Javascripts files
   |    |
   |    |-- templates/
   |    |    |
   |    |    |-- includes/                 # Page chunks, components
   |    |    |    |
   |    |    |    |-- navigation.html      # Top bar
   |    |    |    |-- sidebar.html         # Left sidebar
   |    |    |    |-- scripts.html         # JS scripts common to all pages
   |    |    |    |-- footer.html          # The common footer
   |    |    |
   |    |    |-- layouts/                  # App Layouts (the master pages)
   |    |    |    |
   |    |    |    |-- base.html            # Used by common pages like index, UI
   |    |    |    |-- base-fullscreen.html # Used by auth pages (login, register)
   |    |    |
   |    |  index.html                      # The default page
   |    |  login.html                      # Auth Login Page
   |    |  register.html                   # Auth Registration Page
   |    |  page-404.html                   # Error 404 page (page not found)
   |    |  page-500.html                   # Error 500 page (server error)
   |    |    *.html                        # All other pages provided by the UI Kit
   |
   |-- requirements.txt
   |
   |-- run.py
   |
   |-- ************************************************************************
```

<br />

## Recompile CSS

To recompile SCSS files, follow this setup:

<br />

**Step #1** - Install tools

- [NodeJS](https://nodejs.org/en/) 12.x or higher
- [Gulp](https://gulpjs.com/) - globally 
    - `npm install -g gulp-cli`
- [Yarn](https://yarnpkg.com/) (optional) 

<br />

**Step #2** - Change the working directory to `assets` folder

```bash
$ cd app/base/static/assets
```

<br />

**Step #3** - Install modules (this will create a classic `node_modules` directory)

```bash
$ npm install
// OR
$ yarn
```

<br />

**Step #4** - Edit & Recompile SCSS files 

```bash
$ gulp
```

The generated files (css, min.css) are saved in `static/assets/css` directory.

<br /> 

## Deployment

The project comes with a basic configuration for [Docker](https://www.docker.com/), [HEROKU](https://www.heroku.com/), [Gunicorn](https://gunicorn.org/), and [Waitress](https://docs.pylonsproject.org/projects/waitress/en/stable/).

### [Heroku](https://www.heroku.com/)
---

Steps to deploy on **Heroku**

- [Create a FREE account](https://signup.heroku.com/) on Heroku platform
- [Install the Heroku CLI](https://devcenter.heroku.com/articles/getting-started-with-python#set-up) that match your OS: Mac, Unix or Windows
- Open a terminal window and authenticate via `heroku login` command
- Clone the sources and push the project for LIVE deployment

```bash
$ # Clone the source code:
$ git clone https://github.com/app-generator/jinja-volt-dashboard.git
$ cd jinja-volt-dashboard
$
$ # Check Heroku CLI is installed
$ heroku -v
heroku/7.25.0 win32-x64 node-v12.13.0 # <-- All good
$
$ # Check Heroku CLI is installed
$ heroku login
$ # this commaond will open a browser window - click the login button (in browser)
$
$ # Create the Heroku project
$ heroku create
$
$ # Trigger the LIVE deploy
$ git push heroku master
$
$ # Open the LIVE app in browser
$ heroku open
```

<br />

### [Gunicorn](https://gunicorn.org/)
---

Gunicorn 'Green Unicorn' is a Python WSGI HTTP Server for UNIX.

> Install using pip

```bash
$ pip install gunicorn
```
> Start the app using gunicorn binary

```bash
$ gunicorn --bind 0.0.0.0:8001 run:app
Serving on http://localhost:8001
```

Visit `http://localhost:8001` in your browser. The app should be up & running.

<br />

### [Waitress](https://docs.pylonsproject.org/projects/waitress/en/stable/)
---

Waitress (Gunicorn equivalent for Windows) is meant to be a production-quality pure-Python WSGI server with very acceptable performance. It has no dependencies except ones that live in the Python standard library.

> Install using pip

```bash
$ pip install waitress
```
> Start the app using [waitress-serve](https://docs.pylonsproject.org/projects/waitress/en/stable/runner.html)

```bash
$ waitress-serve --port=8001 run:app
Serving on http://localhost:8001
```

Visit `http://localhost:8001` in your browser. The app should be up & running.

<br />

## Credits & Links

- [Flask Framework](https://www.palletsprojects.com/p/flask/) - The official website

<br />

---
[Flask DataTables Sample](https://blog.appseed.us/flask-data-tables-open-source-sample/) - Provided by **AppSeed** [App Generator](https://appseed.us/app-generator).


CHANGELOG.md
# Change Log

## [0.0.5] 2022-02-07
### Improvements

- Toast notification for `inline` actions
  - Cell Edit
  - Row deletion

## [0.0.4] 2022-02-07
### Improvements

- Added `inline edit` for rows

## [0.0.3] 2022-02-07
### Improvements

- New API Methods 
  - CRUD methods over `Data` table 

## [0.0.2] 2022-02-05
### Improvements

- Added API Node
- New DataTables implementations:
 - via Controller
 - via API
 - loaded from disk 

## [0.0.1] 2022-02-05
### Minimal Version

- `Simple-DataTables`: **v3.2.0**


run.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

from app import app, db
from api.models import Data
from flask_migrate import Migrate

import csv
from datetime import datetime
from random   import randint

# Migrate APP (if necessary)
Migrate(app, db)

# Inner command 
def load_cmd(aUseRandomValues=None):

    if aUseRandomValues:
        print( ' >>> Randomize Values ')
    else:
        print( ' >>> Use Values from input file ')

    # Create Tables (if not exist)
    db.create_all()

    # Truncate 
    db.session.query(Data).delete()
    db.session.commit()

    with open('media/data.csv', newline='') as csvfile:
        
        csvreader = csv.reader(csvfile) # load file
        header    = next(csvreader)     # ignore header (1st line)
        
        '''
        Expected format: 
            HEADER: product_code,product_info,value,currency,type
            SAMPLE: Lenovo_Ideapad_3i, Lenovo Ideapad 3i 14.0inch FHD Laptop,9.5,euro,transaction
            
            product_code (string)  : Lenovo_Ideapad_3i
            product_info (string)  : Lenovo Ideapad 3i 14.0inch FHD Laptop
            value        (integer) : 9 
            currency     (string)  : usd, eur 
            type         (string)  : transaction (hardoded)
        '''

        iter = 0 # used for timestamp
        for row in csvreader:

            iter += 1

            if len( row ) != 5:
                print( ' >>> Error parsing line ('+str(iter)+') -> ' + ' '.join([str(elem) for elem in row]) )
                continue        

            item_code     = row[0]
            item_name     = row[1]

            if aUseRandomValues:
                item_value = randint(5, 100)
            else:
                item_value = row[2]

            item_currency = row[3]
            item_type     = row[4]

            # randomize in the past the transaction date
            # The distribuition range ~1 month 
            item_ts       = datetime.utcnow().timestamp() - ( 2 * iter * randint(5, 10) * 3600 )
            item_ts       = int( item_ts )

            _data = Data(code=item_code, name=item_name, value=item_value, currency=item_currency, type=item_type, ts=item_ts)

            db.session.add(_data)

        db.session.commit()

@app.cli.command("load_data")
def load_data():

    return load_cmd()

@app.cli.command("load_random_data")
def load_random_data():

    return load_cmd( True )

if __name__ == "__main__":
    app.run()



api/models.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

from datetime import datetime
import json

from app import db

class Data(db.Model):

    __tablename__ = 'data'

    id = db.Column(db.Integer, primary_key=True)

    code     = db.Column(db.String(64))   # product code 
    name     = db.Column(db.String(128))  # product name
    value    = db.Column(db.Integer)      # numeric
    currency = db.Column(db.String(10))   # string: usd, euro
    type     = db.Column(db.String(64))   # transaction
    ts       = db.Column(db.Integer, default=datetime.utcnow().timestamp())

    def __repr__(self):
        return self.toJSON()

    def save(self):
        db.session.add(self)
        db.session.commit()

    def update_value(self, new_value):
        self.value = new_value

    @classmethod
    def get_by_id(cls, id):
        return cls.query.filter_by(id=id).first()

    def toDICT(self):

        cls_dict          = {}
        cls_dict['ID']   = self.id
        cls_dict['Product Code'] = self.code
        cls_dict['Description'] = self.name
        cls_dict['Price'] = self.value
        cls_dict['Currency'] = self.currency
        cls_dict['TimeStamp'] = datetime.utcfromtimestamp(self.ts).strftime('%Y-%m-%d')

        return cls_dict

    def toJSON(self):

        return self.toDICT()


api/routes.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

import os, json
from flask import request, send_from_directory
from flask_restx import Api, Resource, fields

from app import app, db
from api.models import Data
from api import rest_api

from datetime import datetime
from sqlalchemy import desc, asc

"""
API Interface:
   
   - /data
       - GET: return all items
       - POST: create a new item
   
   - /data/:id
       - GET    : get item
       - PUT    : update item
       - DELETE : delete item
"""

"""
Flask-RestX models Request & Response DATA
"""

# Used to validate input data for creation
create_model = rest_api.model('CreateModel', {"code": fields.String(required=True, min_length=2, max_length=64),
                                              "name": fields.String(required=True, min_length=2, max_length=128),
                                              "value": fields.Integer(required=True),
                                              "currency": fields.String(required=True, min_length=3, max_length=3),
                                              "type": fields.String(required=False, default='transaction'),
                                            })

# Used to validate input data for update
update_model = rest_api.model('UpdateModel', {"code": fields.String(required=False, min_length=2, max_length=64),
                                              "name": fields.String(required=False, min_length=2, max_length=128),
                                              "value": fields.Integer(required=False),
                                              "currency": fields.String(required=False, min_length=3, max_length=3),
                                              "type": fields.String(required=False, default='transaction'),
                                            })

# Used to validate input data for single field supdate
update_field_model = rest_api.model('UpdateFieldModel', {"data_name" : fields.String(required=True, min_length=1, max_length=128),
                                                         "data_value": fields.String(required=True, min_length=1, max_length=128)
                                            })

"""
    Return Files - Served by @APP object
"""

@app.route('/api/from_file')
def api_from_file(): 
    return send_from_directory(os.path.join(app.root_path, 'static', 'datatables'), 'data.json')

"""
    Flask-Restx routes
"""

"""
class Data(db.Model):


    id = db.Column(db.Integer, primary_key=True)

    code     = db.Column(db.String(64))   # product code 
    name     = db.Column(db.String(128))  # product name
    value    = db.Column(db.Integer)      # numeric
    currency = db.Column(db.String(10))   # string: usd, euro
    type     = db.Column(db.String(64))   # transaction

    ts       = db.Column(db.Integer, default=datetime.utcnow().timestamp())

"""

@rest_api.route('/api/data')
class Items(Resource):

    """
       Return all items
    """
    def get(self):

        data = []
        
        for item in Data.query.all():
            data.append( item.toDICT() ) 

        response = app.response_class(
            response=json.dumps(data),
            status=200,
            mimetype='application/json'
        )

        return response

    """
       Create new item
    """
    @rest_api.expect(create_model, validate=True)
    def post(self):

        # Read ALL input  
        req_data = request.get_json()

        code     = req_data.get("code")
        name     = req_data.get("name")
        value    = req_data.get("value")
        currency = req_data.get("currency")
        type     = req_data.get("type")

        # Create new object
        new_item = Data(code=code, name=name, value=value, currency=currency, type=type)

        # Save the data
        new_item.save()
        
        response = app.response_class(
            response=json.dumps( new_item.toDICT() ),
            status=200,
            mimetype='application/json'
        )

        return response

@rest_api.route('/api/data/<int:id>')
class ItemManager(Resource):

    """
       Return Item
    """
    def get(self, id):

        response_data = []
        status = 400

        item = Data.get_by_id(id)

        if item:
            status = 200
            response_data = item.toDICT()

        response = app.response_class(
            response=json.dumps( response_data ), 
            status=status,
            mimetype='application/json'
        )

        return response

    """
       Update Item
    """
    @rest_api.expect(update_model, validate=True)
    def put(self, id):

        response_data = []
        status = 400

        # Read ALL input  
        req_data = request.get_json()

        item = Data.get_by_id(id)

        if item:
            
            status = 200
            
            # Update data
            if req_data.get("code"):
                item.code = req_data.get("code")
            
            if req_data.get("name"):
                item.name = req_data.get("name")
            
            if req_data.get("value"):
                item.value = req_data.get("value")
            
            if req_data.get("currency"):
                item.currency = req_data.get("currency")
            
            if req_data.get("type"):
                item.type = req_data.get("type")

            # Save the data
            item.save()

            response_data = item.toDICT()

        response = app.response_class(
            response=json.dumps( response_data ), 
            status=status,
            mimetype='application/json'
        )

        return response

    """
       Delete Item
    """
    def delete(self, id):

        response_data = []
        status = 400

        item = Data.get_by_id(id)

        if item:
            
            status = 200

            # Delete and save the change
            Data.query.filter_by(id=id).delete()
            db.session.commit()

        response = app.response_class(
            response=json.dumps( response_data ), 
            status=status,
            mimetype='application/json'
        )

        return response

@rest_api.route('/api/data/field/<int:id>')
class FieldManager(Resource):

    """
       Return Item
    """
    @rest_api.expect(update_field_model, validate=True)
    def put(self, id):

        response_data = []
        status = 400

        # Read ALL input  
        req_data = request.get_json()

        item = Data.get_by_id(id)

        if item:
            
            status = 200
            
            data_name  = req_data.get("data_name"  )
            data_value = req_data.get("data_value" )
            
            if data_name == 'code':
                item.code = data_value        

            if data_name == 'name':
                item.name = data_value        

            # Save the data
            item.save()

            response_data = item.toDICT()

        response = app.response_class(
            response=json.dumps( response_data ), 
            status=status,
            mimetype='application/json'
        )

        return response     

app/static/datatables/app.js
/*

Copyright (c) 2019 - present AppSeed.us

*/

"use strict";
const doc = document;
doc.addEventListener("DOMContentLoaded", function(event) {

    // Used for API Requests 
    var xhr1 = new XMLHttpRequest();
 
    // Toasts
    const notyf = new Notyf({
        position: {
            x: 'right',
            y: 'top',
        },
        types: [
            {
                type: 'info',
                background: '#FA5252',
                icon: {
                    className: 'fas fa-comment-dots',
                    tagName: 'span',
                    color: '#fff'
                },
                dismissible: false
            }
        ]
    });

    // Used by cell edit
    document.addEventListener('keydown', event => {

        // Catch 'ENTER' event
        if (event.key === 'Enter') {
          
            // If applies on a TD  
            if (event.target.matches('td.editable')) {
          
                // Drop the default event
                event.preventDefault();
                            
                let td = event.target;
                let tr = event.target.closest('tr.editable');

                let data_id    = tr.dataset.id;            
                let data_name  = td.dataset.name;
                let data_value = td.textContent;
                
                xhr1.open("PUT", "/api/data/field/" + data_id); 
                xhr1.setRequestHeader("Content-Type", "application/json");
                
                // Check Status
                xhr1.onreadystatechange = function() {
                    
                    if (this.status == 200 && this.readyState == 4) { 

                        notyf.open({
                            type: 'success',
                            message: 'Information saved successfully'
                        });                        
                    }    
                    
                    // XMLHttpRequest returns 0 on success
                    if ( (this.status > 0) && (this.status != 200) ) { 
                        
                        notyf.open({
                            type: 'error',
                            message: 'Error!'
                        });
                    }

                };//end onreadystate

                xhr1.send(JSON.stringify( { data_name: data_name, data_value: data_value } ));

                // Disable 'EDITABLE' property    
                td.setAttribute("contenteditable", "false");

            } // END if (event.target.matches('td.editable')) {

        } // END if (event.key === 'Enter') { 

      });

    // handle Clicks     
    document.addEventListener('click', function (event) {

        // Cell Edit event
        if (event.target.matches('td.editable')) {

            // Drop the default event
            event.preventDefault();
        
            let td = event.target;
            let tr = event.target.closest('tr.editable');

            let data_id   = tr.dataset.id;            
            let data_name = td.dataset.name;
                        
            // Enable 'EDITABLE' property
            td.setAttribute("contenteditable", "true");
        }

        // ROW Delete event
        if (event.target.matches('.row-delete')) {

            // Drop the default event
            event.preventDefault();
        
            let tr = event.target.closest('tr.editable');
            let data_id = tr.dataset.id;

            xhr1.open('DELETE', "/api/data/" + data_id, true);
            
            // Check Status
            xhr1.onreadystatechange = function() {
                
                if (this.status == 200 && this.readyState == 4) { 

                    tr.remove();

                    notyf.open({
                        type: 'success',
                        message: 'Information deleted successfully'
                    });     
                }    

                // XMLHttpRequest returns 0 on success
                if ( (this.status > 0) && (this.status != 200) ) { 

                    notyf.open({
                        type: 'error',
                        message: 'Error! ' + this.status
                    }); 
                }

            };//end onreadystate

            xhr1.send();

        }
    
    }, false); 

}); 


app/views.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

# Python modules
import os, logging 
from datetime import datetime

# Flask modules
from flask   import render_template, request, send_from_directory
from jinja2  import TemplateNotFound
from api.models import Data

# App modules
from app import app

'''
DATA tables routes
'''

# DataTables processed in controler
@app.route('/datatables/')
def datatables():

    # Expected Data
    '''
    <table class="table">
        <thead>
            <tr>
                <th>Name</th>
                <th>Ext.</th>
                <th>City</th>
                <th data-type="date" data-format="YYYY/DD/MM">Start Date</th>
                <th>Completion</th>
            </tr>
        </thead>
        <tbody>
            <tr><td>Unity Pugh</td><td>9958</td><td>Curicó</td><td>2005/02/11</td><td>37%</td></tr>
            <tr><td>Theodore Duran</td><td>8971</td><td>Dhanbad</td><td>1999/04/07</td><td>97%</td></tr>
            <tr><td>Kylie Bishop</td><td>3147</td><td>Norman</td><td>2005/09/08</td><td>63%</td></tr>
            <tr><td>Willow Gilliam</td><td>3497</td><td>Amqui</td><td>2009/29/11</td><td>30%</td></tr>
            <tr><td>Blossom Dickerson</td><td>5018</td><td>Kempten</td><td>2006/11/09</td><td>17%</td></tr>
        </tbody>
    </table>
    '''
    
    '''
    Data:
        id = db.Column(db.Integer, primary_key=True)

        code     = db.Column(db.String(64))   # product code 
        name     = db.Column(db.String(128))  # product name
        value    = db.Column(db.Integer)      # numeric
        currency = db.Column(db.String(10))   # string: usd, euro
        type     = db.Column(db.String(64))   # transaction
        ts       = db.Column(db.Integer, default=datetime.utcnow().timestamp())    
    '''

    table  = ''
    table += '<table class="table hoverTable">'
    table += '    <thead>'
    table += '        <tr>'
    table += '            <th>ID</th>'
    table += '            <th>Product Code</th>'
    table += '            <th>Description</th>'
    table += '            <th>Price</th>'
    table += '            <th>Currency</th>'
    table += '            <th>Timestamp</th>'
    table += '            <th>&nbsp;</th>'
    table += '        </tr>'
    table += '    </thead>'
    table += '    <tbody>'
    
    # Sample
    # table += '        <tr><td>1</td><td>PRD_CODE</td><td>PRD_INFO</td><td>99</td><td>usd</td><td>2022-02-03</td></tr>'
    
    for item in Data.query.all():    

        # Format the date
        ts = datetime.utcfromtimestamp(item.ts).strftime('%Y-%m-%d')
       #table += '        <tr><td>{0}</td><td>{1}</td><td>{2}</td><td>{3}</td><td>{4}</td><td>{5}</td></tr>'.format( item.id, item.code, item.name, item.value, item.currency, ts )
        
        # Open ROW
        table += '<tr data-id="' + str( item.id   ) + '" class="editable">'
        
        table += '<td data-name="id">' + str( item.id   )  + '</td>' # ID
        table += '<td class="editable" data-name="code">' + item.code         + '</td>' # ID
        table += '<td class="editable" data-name="name">' + item.name         + '</td>' # ID
        table += '<td class="editable" data-name="value">' + str( item.value ) + '</td>' # ID
        table += '<td class="editable" data-name="currency">' + item.currency     + '</td>' # ID
        table += '<td data-name="ts">' + ts                + '</td>' # ID

        table += '<td><a href="#" class="row-delete text-danger me-3">delete</a></td>' # ID

        # Close ROW
        table += '</tr>'

    table += '    </tbody>'
    table += '</table>' 
    
    return render_template( 'datatables/datatables.html', segment='datatables', table=table )

# DataTables lazy loaded via API
@app.route('/datatables-api/')
def datatables_api():

    return render_template( 'datatables/datatables-via-api.html', segment='datatables-api' )

# DataTables lazy loaded via API
@app.route('/datatables-from-file/')
def datatables_from_file():

    return render_template( 'datatables/datatables-via-file.html', segment='datatables-from-file' )

# App main route + generic routing
@app.route('/')
def index():

    return render_template( 'index.html', segment='index.html' )

# App main route + generic routing
@app.route('/<path>')
def pages(path):

    try:

        # Detect the current page
        segment = get_segment( request )

        # Serve the file (if exists) from app/templates/FILE.html
        return render_template( 'pages/' + path, segment=segment )
    
    except TemplateNotFound:
        return render_template('pages/page-404.html'), 404

def get_segment( request ): 

    try:

        segment = request.path.split('/')[-1]

        if segment == '':
            segment = 'index'

        return segment    

    except:
        return None  
